portfolio/src/main/java/com/google/sps/servlets/SpotifyServlet.java

```diff
@@ -0,0 +1,150 @@
+    Integer seconds = (Integer)(milliseconds/ 1000) % 60;
+    Integer minutes = (Integer)(milliseconds/ (1000) / 60);
+    return String.format("%d:%02d", minutes, seconds);
+  }
+
+  public static void authorizationCode() {
+    try {
+      final AuthorizationCodeCredentials authorizationCodeCredentials = authorizationCodeRequest.execute();
+      spotifyApi.setAccessToken(authorizationCodeCredentials.getAccessToken());
+      spotifyApi.setRefreshToken(authorizationCodeCredentials.getRefreshToken());
+      System.out.println("Expires in: " + authorizationCodeCredentials.getExpiresIn());
+    } catch (IOException | SpotifyWebApiException | ParseException e) {
+      System.out.println("Error: " + e.getMessage());
+    }
+  }
+
+  public static void authorizationCodeRefresh() {
+    try {
+      final AuthorizationCodeCredentials authorizationCodeCredentials = authorizationCodeRefreshRequest.execute();
+ 
+      spotifyApi.setAccessToken(authorizationCodeCredentials.getAccessToken());
+      spotifyApi.setRefreshToken(authorizationCodeCredentials.getRefreshToken());
+ 
+      System.out.println("Expires in: " + authorizationCodeCredentials.getExpiresIn());
+    } catch (IOException | SpotifyWebApiException | ParseException e) {
+      System.out.println("Error: " + e.getMessage());
+    }
+  }
+
+  public static String getCurrentTrack() {
+    try {
+      final CompletableFuture<CurrentlyPlaying> currentlyPlayingFuture = getUsersCurrentlyPlayingTrackRequest.executeAsync();
+
+      final CurrentlyPlaying currentlyPlaying = currentlyPlayingFuture.join();
+
+      if (currentlyPlaying == null) {
+        return "No song currently playing.";
+      }
+
+      IPlaylistItem item = currentlyPlaying.getItem();
+      String songName = "";
+      String artists = "";
+      
+      if (item instanceof Track) {
+        songName = ((Track)item).getName();
+
+        for (int iterator = 0; iterator < ((Track)item).getArtists().length; iterator++) {
+          if (iterator == ((Track)item).getArtists().length - 1) {
+            artists += ((Track)item).getArtists()[iterator].getName();
+            break;
+          }
+          artists += ((Track)item).getArtists()[iterator].getName() + ", ";
+        }
+      }
+
+      Integer progressMilliseconds = currentlyPlaying.getProgress_ms();
+      Integer durationMilliseconds = ((Track)item).getDurationMs();
+
+      return (songName + " by " + artists + " at " +
+              formatAsMinutesAndSeconds(progressMilliseconds) +
+              " of " + formatAsMinutesAndSeconds(durationMilliseconds));
+    } catch (CompletionException e) {
+      return "No song currenly playing.";
+    } catch (CancellationException e) {
+      return "Error receiving currenly playing song.";
+    }
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    if (oauthToken.isEmpty() && request.getParameter("code") == null) {
```

Comment:
Hmm, does this mean you have to personally visit your site every once in awhile to refresh the oauth token?

Reply:
I realized that I forgot a catch statement in getCurrentTrack() which would handle the authorization code refresh if an error occurred. Since the JS function that fetches /spotify is on load, I personally wouldn't have to visit the site myself to refresh the oauth token because it would get refreshed anytime someone else visits if it's expired.